fix a lot of issues related to no sound on mac

anatoly-os opened the pull request:
If we don't specify framesPerBuffer parameter in Pa_OpenStream, PortAudio will choose (randomly?) the value. In Voice::write, we assume that RestN which initially keeps the value of frames from pa_callback will be bigger than env_data->count. The last is 43 on my Mac (47 on my Ubuntu machine), but default frames count from pa callback on failed MacBook 12" is 39. So, we cannot correctly process all frames and envelop and produce sound, specifically amplitude calculation fails and Voice::write returns under this if statement 'if (volenv_section == FLUID_VOICE_ENVDELAY)'.

So, I hardcoded 630 since it is the value from my Mac. Also I've got 1027 for Ubuntu.

anatoly-os commented:
Let's open a discussion. I've fixed the bug, sound appeared on MacBook 12".

- But I'm not sure neither about hardcoded value nor that we should modify framesPerBuffer, but leave envelope calculations.

- I don't know why port audio on that MacBook processes only about 30 frames, not hundreds.

- I'm also not aware about performance issues which we can face with after specify frames directly.

Jojo-Schmitz commented:
If only Mac is affected by that problem, why making that change for all OSs?
Esp. as just a few lines above we have a Mac specific change already?

anatoly-os commented:
I'm not sure about this. We know about Mac issues, also we know about Bluetooth issues, which may be related. But, if we don't know, it doesn't mean that no issues exist. Since we have specific limit in our code, we cannot expect any value here. So, we have to control it somehow. I would prefer to set minimal value, but there is no option.

Jojo-Schmitz commented:
AFAIK the bluetooth issues are Mac only too?

Jojo-Schmitz commented on mscore/pa.cpp:
```diff
@@ -125,11 +125,15 @@ bool Portaudio::init(bool)
       out.hostApiSpecificStreamInfo = 0;
 
-      err = Pa_OpenStream(&stream, 0, &out, double(_sampleRate), 0, 0, paCallback, (void*)this);
+      int defaultFramesToProcessFromCallback = 0;
```
`unsigned long` rather than `int`

polarbreeze commented:
This is old but maybe it helps?
http://portaudio.music.columbia.narkive.com/taCfo2RQ/portaudio-osx-i-o-buffer-size

lasconic commented:
Awesome. Very glad we have a solid explanation for the problem. 

I don't think it's a good idea to hard code the `framesPerBuffer` value. It's discouraged by PortAudio. The problem has been introduced in Fluid... it needs to be fixed in Fluid...

I'm also pretty sure that it's related with "Bluetooth not working" I believe bluetooth will typically use smaller `framesPerBuffer` to cope with latency.

S

polarbreeze commented:
Is there something to learn from what Audacity does? It is working reliably for me on MacBook 12", for built-in and BT and USB.
Some things I noticed:
1. Audacity separates the input selector from the output selector. Good idea because otherwise it's too easy to select an input device for the output (which results in silence) - in MS the input device has the same name as the output device and they both appear in the device selector menu - so you can't tel which is which. (Maybe the input device should not be presented at all?)
2. In Audacity it's possible to switch back and forth between different output devices without restarting the program, which is easier and less error prone ("did I remember to restart the program?")
3. Portaudio version 19.5.0-devel.

![screen shot 2018-02-25 at 06 08 03](https://user-images.githubusercontent.com/26418826/36640858-d26d9e9c-19f4-11e8-99e4-9b367a59af4e.jpg)

polarbreeze commented:
I'm wondering... if this is actually consistent with the symptoms I am seeing on my MacBook 12":
1. The dropdown are not blank - they are properly populated
2. When "built-in" audio is selected:
a) With both 2.0.3 and 2.1: this console coreaudio error message is repeated every few seconds, even if no score is playing:
"HALS_IOA1Engine.cpp:365:EndWriting: HALS_IOA1Engine::EndWriting: got an error from the kernel trap, Error: 0xE00002EE"
b) With 2.1, there is no audio from the score or from note entry; however, the metronome can be heard normally.
c) With 2.0.3, the audio works properly.
3. When "Bluetooth, "HDMI" or "USB" audio is selected:
a) There is no error message of that sort either with 2.0.3 or with 2.1
b) The audio works properly with both 2.0.3 and 2.1

anatoly-os commented:
Don't afraid huge changes in voice.cpp - it is just extracting huge write logic to separate methods. What I've done actually is generating envelope for minimum 100 frames. If framesBuf from pa is less, we generate envelope for 100 anyway but send only framesBuf number of frames. It results in weak attack and ending sound of notes. I'm trying to use cache data, but still no success.

To emulate no sound case as on MacBook 12 or bluetooth, set fifth argument of Pa_OpenStream in pa.cpp to desired small buff size (39 - for MacBook, 14 - for bluetooth from experiments).

polarbreeze commented:
Hi Anatoly, nice job there!

1. Just to let you know that on MacBook 12” I have no problem with MuseScore Bluetooth audio.
Also, HDMI and USB audio outputs are OK.
Only built-in audio is a problem. (And also AirPlay, but I haven’t rigorously confirmed that one).
Is this consistent with your findings?

2. I have found that if built-in audio is in a working state (rarely!), I can cause it to fail by selecting and deselecting the Bluetooth. However, BT itself does not have an issue - at least I have not observed one yet despite extensive testing.

3. A related issue is that MS includes input devices in the output device dropdown, and they are often not identified as inputs. Therefore users may select an input device by accident, resulting in silence - I think that may be responsible for some of the no-audio reports.

Sorry to bug you, hope it’s useful!
Howard


> On Feb 27, 2018, at 08:15, Anatoly <notifications@github.com> wrote:
> 
> Don't afraid huge changes in voice.cpp - it is just extracting huge write logic to separate methods. What I've done actually is generating envelope for minimum 100 frames. If framesBuf from pa is less, we generate envelope for 100 anyway but send only framesBuf number of frames. It results in weak attack and ending sound of notes. I'm trying to use cache data, but still no success.
> 
> To emulate no sound case as on MacBook 12 or bluetooth, set fifth argument of Pa_OpenStream in pa.cpp to desired small buff size (39 - for MacBook, 14 - for bluetooth from experiments).
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/musescore/MuseScore/pull/3492#issuecomment-368869384>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AZMeiv49__Ra1tQhJ4IxoZ8VdxB5PLfTks5tZAACgaJpZM4SR61U>.
>

anatoly-os commented:
Finished implementation of the cache. Reverted all changes in pa.cpp to keep optimal buffer size. So, for now, I cannot imagine a case when there is no sound because of buffer size. This PR is supposed to fix MacBook 12" no sound issues, bluetooth issues. According to my investigation, slow computers also could be affected.

I tried to describe all the changes in commit log, but logic in Voice and dip generation is too complex. BTW, I enriched my knowledge :)

This PR should be thoroughly tested in terms of sound quality. There should be no glitches, no rattling, no noise. Everything we find, I will fix...

I'm happy now. Long weekends are waiting for me :)

Jojo-Schmitz commented:
Would be great to also fix the display of input devices in the output list and the truncation of the entries in that list

lasconic commented:
*fix the display of input devices in the output list, as selecting those would cause no sound also*

This is a totally different bug. So please, not in this PR.

anatoly-os commented:
@lasconic for sure. I was not even going to :)

polarbreeze commented:
Hi Anatoly, sounds great! I’m ready to test thoroughly. How will I know which build has the fix?
Thanks, Howard

> On Mar 1, 2018, at 12:36, Anatoly <notifications@github.com> wrote:
> 
> Finished implementation of the cache. Reverted all changes in pa.cpp to keep optimal buffer size. So, for now, I cannot imagine a case when there is no sound because of buffer size. This PR is supposed to fix MacBook 12" no sound issues, bluetooth issues. According to my investigation, slow computers also could be affected.
> 
> I tried to describe all the changes in commit log, but logic in Voice and dip generation is too complex. BTW, I enriched my knowledge :)
> 
> This PR should be thoroughly tested in terms of sound quality. There should be no glitches, no rattling, no noise. Everything we find, I will fix...
> 
> I'm happy now. Long weekends are waiting for me :)
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/musescore/MuseScore/pull/3492#issuecomment-369668815>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AZMeisH9mi8NYoe2YpI12xttSvdfJ1dFks5taDFfgaJpZM4SR61U>.
>

anatoly-os commented:
Well, there is no build with fix I suppose until this PR will be merged to master... Are you a developer? Are you building MuseScore sources?

polarbreeze commented:
Hi Anatoly - I’m a software engineer but I’m not a developer on this project. 
For this one I’m just an anxious user and tester!
I’m hoping to get the sound fixed on v 2.1 (or 2.2 I guess).
I’m not familiar with how up-revs happen here…
Howard

> On Mar 1, 2018, at 13:18, Anatoly <notifications@github.com> wrote:
> 
> Well, there is no build with fix I suppose until this PR will be merged to master... Are you a developer? Are you building MuseScore sources?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/musescore/MuseScore/pull/3492#issuecomment-369682320>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AZMeimxTwAx6wxVTug0nPWITO4wTHI57ks5taDuGgaJpZM4SR61U>.
>

anatoly-os commented:
But since Travis and AppVeyor work on this build, I believe they should create build with this PR. @lasconic, am I right?

Jojo-Schmitz commented:
In this case see https://ci.appveyor.com/project/MuseScore/musescore/build/1.0.559/artifacts

lasconic commented:
Only Appveyor creates downloadable builds from PR. So only for windows.

lasconic commented:
The good news, it sounds ok with internal speakers and I have sound via bluetooth (after a MuseScore restart but that's expected)

The not so good news, Bluetooth playback is not as good than the internal speakers, I hear some glitches.

anatoly-os commented:
@lasconic did you use MacOS debug build for testing using code from this PR or downloaded AppVeyor's package and ran MuseScore on virtual machine?

I've checked sound on both Windows debug build and downloaded AppVeyor's build with my JBL Charger 3 Bluetooth speaker. Sound is good, without noise or glitches with the default (1024 even with Bluetooth) number of frames and the hardcoded 14 frames as you mentioned you had for bluetooth headphones, Sound is still great even after switching devices on the fly.

Maybe, it is about computers with less memory or CPU resources...

lasconic commented:
I tested on MacOS debug. Tried now with release. Same problem : glitches with bluetooth headphones.
Internal speakers or wired headphones sound ok.

lasconic commented:
With my QC35 headphones, n = 14 in voice::write()

polarbreeze commented:
Significant clue someone mentioned and I’ve confirmed to be dependable after multiple tests (with released version 2.1) - no counterexample so far:
1. Start with MuseScore in its “no-audio” situation.
2. Close MS
3. Plug in the earbuds
4. Start MS
Result: earbud audio works
5. Unplug the earbuds (with MS still running)
Result: built-in audio works


> On Mar 5, 2018, at 14:49, Anatoly <notifications@github.com> wrote:
> 
> @lasconic <https://github.com/lasconic> did you use MacOS debug build for testing using code from this PR or downloaded AppVeyor's package and ran MuseScore on virtual machine?
> 
> I've checked sound on both Windows debug build and downloaded AppVeyor's build with my JBL Charger 3 Bluetooth speaker. Sound is good, without noise or glitches with the default (1024 even with Bluetooth) number of frames and the hardcoded 14 frames as you mentioned you had for bluetooth headphones, Sound is still great even after switching devices on the fly.
> 
> Maybe, it is about computers with less memory or CPU resources...
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/musescore/MuseScore/pull/3492#issuecomment-370540874>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AZMeilptp8zxkOra8dRAbMFqL-UezOEPks5tbZZXgaJpZM4SR61U>.
>

anatoly-os commented:
@polarbreeze, we are discussing not merged changes to master and 2.2. After I finish with it, you can test it using nightly build.

Speaking about switching devices, the issue exists, indeed. But it is another issue, so I'll switch there later.

polarbreeze commented:
OK, thanks Anatoly

> On Mar 6, 2018, at 04:00, Anatoly <notifications@github.com> wrote:
> 
> @polarbreeze <https://github.com/polarbreeze>, we are discussing not merged changes to master and 2.2. After I finish with it, you can test it using nightly build.
> 
> Speaking about switching devices, the issue exists, indeed. But it is another issue, so I'll switch there later.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/musescore/MuseScore/pull/3492#issuecomment-370710578>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AZMeinIIjSZ1EAuZfKyu-Q_Q80q6ByJuks5tblA9gaJpZM4SR61U>.
>

anatoly-os commented:
@lasconic I've added comments and fixed bluetooth devices issue.

There is last issue with small noise on stopping playback, but it is not critical. I'll find way to fix it.

lasconic commented:
Awesome. Let me test.

lasconic commented:
met's align master with 2.2 by merging this one on master too. See https://github.com/musescore/MuseScore/pull/3518
